core/headerchain.go

```diff
@@ -105,302 +122,212 @@ func NewHeaderChain(chainDb ethdb.Database, config *params.ChainConfig, engine c
-		} else {
-			hash = header.Hash()
+		// Trim the branch before dequeueing
+		commonHeader := hc.findCommonHeader(hc.heads[0])
+		err = hc.trim(commonHeader, hc.heads[0])
+		if err != nil {
+			return err
 		}
 
-		newTd, err = hc.CalcTd(header)
-		if err != nil {
-			return &headerWriteResult{}, errors.New("error calculating the total td for the header")
+		// dequeue
+		hc.heads[0] = nilHeader
+		hc.heads = hc.heads[1:]
+	}
+	// Add to the heads queue
+	hc.heads = append(hc.heads, block.Header())
+
+	// Sort the heads by number
+	sort.Slice(hc.heads, func(i, j int) bool {
+		return hc.heads[i].Number[types.QuaiNetworkContext].Uint64() < hc.heads[j].Number[types.QuaiNetworkContext].Uint64()
+	})
+
+	return nil
+}
+
+// SetCurrentHeader sets the in-memory head header marker of the canonical chan
+// as the given header.
+func (hc *HeaderChain) SetCurrentHeader(head *types.Header) error {
+	hc.currentHeader.Store(head)
+	hc.currentHeaderHash = head.Hash()
+	headHeaderGauge.Update(head.Number[types.QuaiNetworkContext].Int64())
+
+	//Update canonical state db
+	//Find a common header
+	commonHeader := hc.findCommonHeader(head)
+	parent := head
+
+	// Delete each header and rollback state processor until common header
+	// Accumulate the hash slice stack
+	var hashStack []*types.Header
+	for {
+		if parent.Hash() == commonHeader.Hash() {
+			break
 		}
 
-		number := header.Number[types.QuaiNetworkContext].Uint64()
-
-		// If the parent was not present, store it
-		// If the header is already known, skip it, otherwise store
-		alreadyKnown := parentKnown && hc.HasHeader(hash, number)
-		if !alreadyKnown {
-			// Irrelevant of the canonical status, write the TD and header to the database.
-			rawdb.WriteTd(batch, hash, number, newTd)
-			hc.tdCache.Add(hash, newTd)
-
-			rawdb.WriteHeader(batch, header)
-			inserted = append(inserted, numberHash{number, hash})
-			hc.headerCache.Add(hash, header)
-			hc.numberCache.Add(hash, number)
-			if firstInserted < 0 {
-				firstInserted = i
-			}
+		// Delete the header and the block
+		rawdb.DeleteCanonicalHash(hc.headerDb, parent.Number64())
+
+		//TODO: Run state processor to rollback state
+
+		// Add to the stack
+		hashStack = append(hashStack, parent)
+		parent = hc.GetHeader(parent.Parent(), parent.Number64()-1)
+
+		if parent == nil {
+			log.Warn("unable to trim blockchain state, one of trimmed blocks not found")
+			return nil
 		}
-		parentKnown = alreadyKnown
-		lastHeader, lastHash, lastNumber = header, hash, number
 	}
 
-	// Skip the slow disk write of all headers if interrupted.
-	if hc.procInterrupt() {
-		log.Debug("Premature abort during headers import")
-		return &headerWriteResult{}, errors.New("aborted")
+	// Run through the hash stack to update canonicalHash and forward state processor
+	for i := len(hashStack); i > 0; i-- {
+		rawdb.WriteCanonicalHash(hc.headerDb, hashStack[i].Hash(), hashStack[i].Number64())
+
+		//TODO: Run the state processor forward
```

Comment:
It looks like state processing actually happens in `bc.Append()`, independent of this function. Is it meant to exist in two places? Without state processing here (and without state snapshotting), I'm not sure the EVM survives a reorg right now...